From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luke Chambers <consolelogluke@gmail.com>
Date: Wed, 26 Apr 2023 22:15:36 -0400
Subject: [PATCH] Add registry client


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index a68ccce5be03e847116519e152945f88dab71d3c..7e348bd0cffaecff53280391b3e5de8eb2030795 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -295,6 +295,7 @@ public final class CraftServer implements Server {
     public static Exception excessiveVelEx; // Paper - Velocity warnings
     private final io.papermc.paper.logging.SysoutCatcher sysoutCatcher = new io.papermc.paper.logging.SysoutCatcher(); // Paper
     private final CraftPotionBrewer potionBrewer = new CraftPotionBrewer(); // Paper
+    private final sh.lpx.cardstock.registry.RegistryClient registryClient; // Cardstock
 
     static {
         ConfigurationSerialization.registerClass(CraftOfflinePlayer.class);
@@ -315,6 +316,20 @@ public final class CraftServer implements Server {
         this.serverVersion = CraftServer.class.getPackage().getImplementationVersion();
         this.structureManager = new CraftStructureManager(console.getStructureManager());
 
+        // Cardstock start
+        sh.lpx.cardstock.registry.packet.client.ClientHandshakePacket handshake =
+            new sh.lpx.cardstock.registry.packet.client.ClientHandshakePacket(this.getVersion());
+        try {
+            this.registryClient = sh.lpx.cardstock.registry.RegistryClient.connect("127.0.0.1:15656", handshake);
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to connect to the registry server.", e);
+        }
+
+        Thread registryClientThread = new Thread(this.registryClient::run, "registry-client");
+        registryClientThread.setDaemon(true);
+        registryClientThread.start();
+        // Cardstock end
+
         Bukkit.setServer(this);
 
         // Register all the Enchantments and PotionTypes now so we can stop new registration immediately after
